Allow ref struct to implement interfaces (#7625)
dotnet/csharplang: proposals/ref-struct-interfaces.md

Ref Struct Interfaces
=====

## Summary
This proposal will expand the capabilities of `ref struct` such that they can implement interfaces and participate as generic type arguments.

## Motivation

> The inability for `ref struct` to implement interfaces means they cannot partipcate in fairly fundamental abstraction techniques of .NET. A `Span<T>`, even though it has all the attributes of a sequential list cannot participate in methods that take `IReadOnlyList<T>`, `IEnumerable<T>`, etc ... Instead specific methods must be coded for `Span<T>` that have virtually the same implementation. Allowing `ref struct` to implement interfaces will allow operations to be abstracted over them as they are for other types. 

Reviewer: ```suggestion
The inability for `ref struct` to implement interfaces means they cannot participate in fairly fundamental abstraction techniques of .NET. A `Span<T>`, even though it has all the attributes of a sequential list cannot participate in methods that take `IReadOnlyList<T>`, `IEnumerable<T>`, etc. ... Instead specific methods must be coded for `Span<T>` that have virtually the same implementation. Allowing `ref struct` to implement interfaces will allow operations to be abstracted over them as they are for other types. 
```

## Detailed Design

### ref struct interfaces
<a name="ref-struct-interface"></a>
The language will allow for `ref struct` types to implement interfaces. The syntax and rules are the same as for normal `struct` with a few exceptions to account for the limitations of `ref struct` types. 

The ability to implement interfaces does not impact the existing limitations against boxing `ref struct` instances. That means even if a `ref struct` implements a particular interface,  it cannot be directly cast to it as that represents a boxing action.

```csharp
ref struct File : IDisposable
{
    private SafeHandle _handle;
    public void Dispose()
    {
        _handle.Dispose();
    }
}

File f = ...;
// Error: cannot box `ref struct` type `File`
IDisposable d = f;
```

The ability to implement interfaces is only useful when combined with the ability for `ref struct` to participate in generic arguments (as [laid out later](#ref-struct-generic)).

> To allow for interfaces to cover the full expressiveness of a `ref struct`, the language will allow `[UnscopedRef]` to appear on interface methods and properties. When a `ref struct` member implements an interface member with a `[UnscopedRef]` attribute, that `ref struct` member must also be decorated with `[UnscopedRef]`. The attribute is ignored when the `class` implements the interface. 

Reviewer 2: ```suggestion
To allow for interfaces to cover the full expressiveness of a `ref struct`, the language will allow `[UnscopedRef]` to appear on interface methods and properties. When a `ref struct` member implements an interface member with a `[UnscopedRef]` attribute, that `ref struct` member must also be decorated with `[UnscopedRef]`. The attribute is ignored when a `class` implements the interface. 
```

> To allow for interfaces to cover the full expressiveness of a `ref struct`, the language will allow `[UnscopedRef]` to appear on interface methods and properties. When a `ref struct` member implements an interface member with a `[UnscopedRef]` attribute, that `ref struct` member must also be decorated with `[UnscopedRef]`. The attribute is ignored when the a `class` implements the interface. 

Reviewer 3: ```suggestion
To allow for interfaces to cover the full expressiveness of a `ref struct`, the language will allow `[UnscopedRef]` to appear on interface methods and properties. When a `ref struct` member implements an interface member with a `[UnscopedRef]` attribute, that `ref struct` member must also be decorated with `[UnscopedRef]`. The attribute is ignored when a `class` implements the interface. 
```

> To allow for interfaces to cover the full expressiveness of a `ref struct`, the language will allow `[UnscopedRef]` to appear on interface methods and properties. When a `ref struct` member implements an interface member with a `[UnscopedRef]` attribute, that `ref struct` member must also be decorated with `[UnscopedRef]`. The attribute is ignored when the a `class` implements the interface. 

Reviewer: ```suggestion
To allow for interfaces to cover the full expressiveness of a `ref struct`, the language will allow `[UnscopedRef]` to appear on interface methods and properties. When a `ref struct` member implements an interface member with a `[UnscopedRef]` attribute, that `ref struct` member must also be decorated with `[UnscopedRef]`. The attribute is ignored when the `class` implements the interface. 
```

Default interface methods pose a problem for `ref struct` as there are no protections against the default implementation boxing the `this` member. 

```csharp
interface I1
{
    void M()
    {
        // Error: both of these box if I1 is implemented by a ref struct

>         I1 local1 = this;

Reviewer 4: The runtime will demand a `this` pointer as a boxed interface even without any of these explicit usages. Simply calling this method will break the runtime. So you're right that this needs a runtime exception.

        object local2 = this;
    }
}

ref struct S = I1 { }
```

To handle this a `ref struct` will be forced to implement all members of an interface, even if they have default implementations. The runtime will also be updated to throw an exception if a default interface member is called on a `ref struct` type.

Detailed Notes:
- A `ref struct` can implement an interface
- A `ref struct` cannot participate in default interface members
- A `ref struct` cannot be cast to interfaces it implements as that is a boxing operation

### ref struct Generic Parameters
<a name="ref-struct-generic"></a>
The language will allow for generic parameters to opt into supporting `ref struct` as arguments by using the `allow T : ref struct` syntax: 

```csharp
T Identity<T>(T p) 

>     allow T : ref struct

Reviewer 4: \*standard handwaves around syntax\*

    => p;

// Okay

> Span<T> local = Identity(new Span<int>(new int[10]));

Reviewer: ```suggestion
Span<int> local = Identity(new Span<int>(new int[10]));
```

```

> This is similar to a `where` in that it specifies the capabilities of the generic parameter. The difference is `where` limits the set of types that can fullfill a generic parameter while the behavior defined here expands the set of types. This is effectively an anti-constraint as it removes the implicit constraint that `ref struct` cannot satisfy a generic parameter. As such this is given a new syntax, `allow`, to make that clearer.

Reviewer: ```suggestion
This is similar to a `where` in that it specifies the capabilities of the generic parameter. The difference is `where` limits the set of types that can fulfill a generic parameter while the behavior defined here expands the set of types. This is effectively an anti-constraint as it removes the implicit constraint that `ref struct` cannot satisfy a generic parameter. As such this is given a new syntax, `allow`, to make that clearer.
```

Author: I thought that was wrong but VS code didn't highlight it so I went along with it.

A type parameter bound by `allow T: ref struct` has all of the behaviors of a `ref struct` type: 

1. Instances of it cannot be boxed

> 2. Instances participate in lifetime rules like a normal `ref struct`

Reviewer 4: This is perfectly fine, but I also think we could slip some of this support to a second release if necessary. If we don't allow scoped/unscoped/etc then I think the default ref-struct rules are very useful and the cases where you need different lifetimes are pretty narrow.

Author: Not sure what you want to slip here. Instances of the type parameter need to behave like `ref struct` from a lifetime perspective else it's easy to create memory safety issues. That support is fairly easy to add. Conceptually it's just applying lifetime rules to any type which could be a `ref struct` but really nothing else changes.

Reviewer 4: > Not sure what you want to slip here

Just the lifetime modifier syntax like `scoped`. Agreed it may not be worth it to skip.

> 3. The type parameter cannot be used in `static` fields, elements of an array, etc ...

Reviewer 4: In short, `ref struct` cannot appear on the heap, and the rules are meant to prevent that.

4. Instances can be marked with `scoped`

Examples of these rules in action: 

```csharp
interface I1 { }
I1 M1<T>(T p)
    allow T : ref struct, I1
{
    // Error: cannot box potential ref struct
    return p;
}

> T M2<T>(T p)
> {
>     Span<int> span = stackalloc int[42];
> 
>     // The safe-to-escape of the return is current method because one of the inputs is 
>     // current method
>     T t = M3<T>(span);
> 
>     // Error: the safe-to-escape is current method.
>     return t;
> 
>     // Okay
>     return default;
>     return p;
> }
> 
> T M3<T>(Span<int> span)

Reviewer: Should `M3<T>` have a `Span<T>` parameter? If so, perhaps:

```suggestion
T M2<T>(T p)
    where T : unmanaged
{
    Span<T> span = stackalloc T[42];

    // The safe-to-escape of the return is current method because one of the inputs is 
    // current method
    T t = M3<T>(span);

    // Error: the safe-to-escape is current method.
    return t;

    // Okay
    return default;
    return p;
}

T M3<T>(Span<T> span)
```

Author: This should be `allow T : ref struct`

```csharp
// Proposed
void M<T>(T p)
    where T : IDisposable
    allow T : ref struct
{
    p.Dispose();
}

// Alternative
void M<T>(T p)
    where T : IDisposable, ~ref struct
{
    p.Dispose();
}
```

### Co and contra variance
To be maximally useful type parameters that are `allow T : ref struct` must be compatible with generic variance. Specifically it must be legal for a parameter to be both co/contravariant and also `allow T: ref struct`. Lacking that they would not be usable in many of the most popular `delegate` and `interface` types in .NET like `Func<T>`, `Action<T>`, `IEnumerable<T>`, etc ...

Given there is no actual variance when `struct` are involved these should be compatible. There is still some concern that I'm missing deeply generic variance cases. Need to sit down with @agocke to work out if this is truly safe or if there are deeply generic scenarios that need to be worked out.

### Auto-applying to delegate members
**Decision**: do not auto-apply

> For many generic `delegate` members the language could automatically apply `allow T: ref struct` as it's purely an upside change. Consider that for `Func<> / Action<>` style delegetes there is no downside to expanding to allowing `ref struct`. The language can outline rules where it is safe to automatically apply this anti-constraint. This removes the manual process and would speed up the adoption of this feature.

Reviewer: ```suggestion
For many generic `delegate` members the language could automatically apply `allow T: ref struct` as it's purely an upside change. Consider that for `Func<> / Action<>` style delegates there is no downside to expanding to allowing `ref struct`. The language can outline rules where it is safe to automatically apply this anti-constraint. This removes the manual process and would speed up the adoption of this feature.
```

While that is true it can present a problem in multi-targeted scenarios. Code would compile in one target framework but fail in another. This could lead to confusion with customers and result in a desire for a more explicit opt-in.

### Binary breaking change
Adding `allow T: ref struct` to an existing API is not a source breaking change. It is purely expanding the set of allowed types for an API. Need to track down if this is a binary breaking change or not. Unclear if updating the attributes of a generic parameter constitute a binary breaking change.

## Considerations

### Runtime support

> This feature requires several pieces of support from the runtime / libaries team: 

Reviewer: ```suggestion
This feature requires several pieces of support from the runtime / libraries team: 
```

> - Preventing default interface methods from applying to `ref struct`

Reviewer 4: Yeah, this is an important one that will likely need runtime work.

- API in `System.Reflection.Metadata` for encoding the `gpAcceptByRefLike` value.
- Support for generic parameters being a `ref struct`

Most of this support is likely already in place. The general `ref struct` as generic parameter support is already implemented as described [here](https://github.com/dotnet/runtime/blob/main/docs/design/features/byreflike-generics.md). It's possible the DIM implementation already account for `ref struct`. But each of these items needs to be tracked down.

> ### Span<Span<T>>

Reviewer: ```suggestion
### Span&lt;Span&lt;T&gt;&gt;
```

Author: What does this change do? Today it renders as I would expect so not sure what this change brings.

Reviewer: The title is currently rendered as `Span<Span>` where the nested `<T>` is missing.

```csharp
T Identity<T>(T p) 
    allow T : ref struct
    => p;

// Okay
Span<int> local = Identity(new Span<int>(new int[10]));
```

This is similar to a `where` in that it specifies the capabilities of the generic parameter. The difference is `where` limits the set of types that can fulfill a generic parameter while the behavior defined here expands the set of types. This is effectively an anti-constraint as it removes the implicit constraint that `ref struct` cannot satisfy a generic parameter. As such this is given a new syntax, `allow`, to make that clearer.

A type parameter bound by `allow T: ref struct` has all of the behaviors of a `ref struct` type: 

1. Instances of it cannot be boxed
2. Instances participate in lifetime rules like a normal `ref struct`
3. The type parameter cannot be used in `static` fields, elements of an array, etc ...
4. Instances can be marked with `scoped`

Examples of these rules in action: 

```csharp
interface I1 { }
I1 M1<T>(T p)
    allow T : ref struct, I1
{
    // Error: cannot box potential ref struct
    return p;
}

> T M2<T>(T p)
> {
>     Span<int> span = stackalloc int[42];

Reviewer: ```suggestion
T M2<T>(T p)
    where T : unmanaged
{
    Span<T> span = stackalloc T[42];
```

Author: corrected to `allow T : ref struct`

    // The safe-to-escape of the return is current method because one of the inputs is 
    // current method
    T t = M3<T>(span);

    // Error: the safe-to-escape is current method.
    return t;

    // Okay
    return default;
    return p;
}

> T M3<T>(Span<int> span)

Reviewer: ```suggestion
T M3<T>(Span<T> span)
```
